src/vcpkg/portfileprovider.cpp

```diff
@@ -386,7 +386,71 @@ namespace vcpkg::PortFileProvider
             const std::vector<Path> m_overlay_ports;
             mutable std::map<std::string, Optional<SourceControlFileAndLocation>, std::less<>> m_overlay_cache;
         };
-    }
+
+        struct ManifestProviderImpl : IOverlayProvider
+        {
+            ManifestProviderImpl(const Path& manifest_path, std::unique_ptr<SourceControlFile>&& manifest_scf)
+                : m_scf_and_location{std::move(manifest_scf), manifest_path}
+            {
+            }
+
+            virtual Optional<const SourceControlFileAndLocation&> get_control_file(StringView port_name) const override
+            {
+                if (port_name == m_scf_and_location.source_control_file->core_paragraph->name)
+                {
+                    return m_scf_and_location;
+                }
+
+                return nullopt;
+            }
+
+            virtual void load_all_control_files(
+                std::map<std::string, const SourceControlFileAndLocation*>& out) const override
+            {
+                out.emplace(std::piecewise_construct,
+                            std::forward_as_tuple(m_scf_and_location.source_control_file->core_paragraph->name),
+                            std::forward_as_tuple(&m_scf_and_location));
+            }
+
+            SourceControlFileAndLocation m_scf_and_location;
+        };
+
+        struct CombinedProviderImpl : IOverlayProvider
+        {
+            CombinedProviderImpl(std::vector<std::unique_ptr<IOverlayProvider>>&& providers)
+                : m_providers(std::move(providers))
+            {
+            }
+
+            virtual Optional<const SourceControlFileAndLocation&> get_control_file(StringView port_name) const override
+            {
+                Optional<const SourceControlFileAndLocation&> result;
+                for (auto&& provider : m_providers)
+                {
+                    result = provider->get_control_file(port_name);
+                    if (result)
+                    {
+                        break;
+                    }
+                }
+
+                return result;
+            }
+
+            virtual void load_all_control_files(
+                std::map<std::string, const SourceControlFileAndLocation*>& out) const override
+            {
+                auto first = std::make_reverse_iterator(m_providers.end());
```

Comment:
Note that this order is consistent with https://github.com/microsoft/vcpkg-tool/pull/580